merlin/models/tf/models/base.py

```diff
@@ -1025,6 +1028,148 @@ def _set_save_spec(self, inputs, args=None, kwargs=None):
+        if not isinstance(blocks, (list, tuple)):
+            blocks = [blocks]
+        if isinstance(blocks[0], str):
+            blocks_to_freeze = self.find_blocks_by_name(blocks)
+        elif isinstance(blocks[0], Block):
+            blocks_to_freeze = blocks
+        for b in blocks_to_freeze:
+            b.trainable = False
+        self._frozen_blocks.update(blocks_to_freeze)
+
+    def unfreeze_blocks(
+        self,
+        blocks: Union[Sequence[Block], Sequence[str]],
+    ):
+        """Unfreeze all sub-blocks of given blocks recursively"""
+        if not isinstance(blocks, (list, tuple)):
+            blocks = [blocks]
+        if isinstance(blocks[0], Block):
+            blocks_to_unfreeze = set(self.get_sub_blocks(blocks))
+        elif isinstance(blocks[0], str):
+            blocks_to_unfreeze = self.find_blocks_by_name(blocks)
+        for b in blocks_to_unfreeze:
+            if b not in self._frozen_blocks:
+                warnings.warn(
+                    f"Block or sub-block {b} was not frozen when calling unfreeze_block("
+                    f"{blocks})."
+                )
+            else:
+                self._frozen_blocks.remove(b)
+            b.trainable = True
+
+    def unfreeze_all_frozen_blocks(self):
+        """Unfreeze all blocks (including blocks and sub-blocks) of this model recursively"""
+        for b in self._frozen_blocks:
+            b.trainable = True
+        self._frozen_blocks = set()
+
+    def find_blocks_by_name(self, block_names: Sequence[str]) -> List[Block]:
+        """Get blocks by given block_names, return a list of blocks
+        Traverse(Iterate) the model to check each block (sub_block) by BFS"""
+        result_blocks = set()
+        if not isinstance(block_names, (list, tuple)):
+            block_names = [block_names]
+
+        for block in self.blocks:
+            # Traversse all submodule (BFS) except ModelContext
+            deque = collections.deque()
+            if not isinstance(block, ModelContext):
+                deque.append(block)
+            while deque:
+                current_module = deque.popleft()
+                # Already found all blocks
+                if len(block_names) == 0:
+                    break
+                # Found a block
+                if current_module.name in block_names:
+                    result_blocks.add(current_module)
+                    block_names.remove(current_module.name)
+                for sub_module in current_module._flatten_modules(
+                    include_self=False, recursive=False
+                ):
+                    # Filter out modelcontext
+                    if type(sub_module) != ModelContext:
+                        deque.append(sub_module)
+            if len(block_names) > 0:
+                raise ValueError(f"Cannot find block with the name of {block_names}")
+        return list(result_blocks)
+
+    def get_sub_blocks(self, blocks: Sequence[Block]) -> List[Block]:
```

Comment:
I think it would be better to add this method to the utils, and not make it part of the `Model`. This since it seems useful outside the model as well.